jwst/pipeline/calwebb_spec2.py

```diff
@@ -273,7 +273,7 @@ def process_exposure_product(
         # Produce a resampled product, either via resample_spec for
         # "regular" spectra or cube_build for IFU data. No resampled
         # product is produced for time-series modes.
-        if exp_type in ['NRS_FIXEDSLIT', 'NRS_MSASPEC', 'MIR_LRS-FIXEDSLIT']:
+        if exp_type in ['NRS_FIXEDSLIT', 'NRS_MSASPEC', 'MIR_LRS-FIXEDSLIT'] and not tso_mode:
```

Comment:
I'm not sure `tso_mode` will be set for instances like the multi-integration version of a MIR_LRS-FIXEDSLIT product, because technically it's not a TSO mode. So it might be better to explicitly check for the data model type, i.e. skip resample if a `CubeModel` is being processed.

Reply:
FYI:, the `is_tso` check has a check for `CubeModel` there. Some previous issue discussed this, with anytime a `CubeModel` appears as input, it is TSO.   May not be sound, but is the current standard.

Reply:
OK, so if that's the case, then mechanically this should work as desired. But it won't be obvious to someone just taking the code at face value.

Reply:
Agreed on not clear. Not clear to me if this is a documentation issue, or something else. As said, `is_tso` defines that anything `CubeModel` is time series. Which may be correct. But, as @hbushouse said, it may be more appropriate to make the case that both are checked for, since here being a `CubeModel` has other implications. Just because the check may be happening twice is a non-issue. But, if at some point, if `CubeBuild` is not always pinned to TSO, the explicit check will still be necessary.

Reply:
`CubeModel` isn't actually pinned to TSO.  Wouldn't it be better for `resample` to check the dimension of the data array?  That's the issue, isn't it, that `resample` won't process data unless it's 2D?

Reply:
So the consensus is that `is_tso()` is not appropriate for checking if something is 3D CubeModel or similar, despite it doing the check?

The `is_tso()` check further up in the code is actually the only thing making sure that a _rateints file run through this pipeline will actually produce a _calints output.  I used it for consistency's sake.  If it is not appropriate, then it should probably be changed further up in the code as well?  Thoughts?

Reply:
Also, `CubeModel` is always multi-integration, yes?  And whether a multi-integration cube is used for TSO purposes or some other (something like WFC3 DASH mode), shouldn't really matter to this pipeline or the steps in it.  The processing is the same.

Reply:
It seems like in general, that `resample` and `CubeBuild` should define what they can accept, and return an error that's passed up to calwebb_spec2 which is then handled appropriately? Unless this is an expensive operation? That way the extra checking and decision making pipeline code gets called for the outliers, and the lower code is additionaly beefed up in the case it's called directly.

Reply:
Agree that returning useful errors from steps would be be the best.

As you mentioned in discussion before, it would be nice to have a `JwstError` or `StpipeError` class that could be subclassed with useful boilerplate error messages about completion or failure of steps, invalid input to steps, etc.  Since every step handles success or failure in its own way currently, I feel that is more of a infrastructure upgrade issue, not something unique to `resample`.  So if we update how the steps report success or failure, the pipelines calling them need to be modified as well to handle those properly.  So we need consistent reporting.  Currently, only `Spec2Pipeline` does any sort of error handling in terms of success or failure, so that it can exit gracefully.

Agree though, that this should be on our infrastructure to-do list.  👍

Reply:
If it doesn't make sense to change the flow now, it may make the current code more readable to simply condense the `tso_mode` checks. i.e. set the suffix and make the correct calls in one `if is_tso(input):` block. That way everything explicitly done for tso mode is structured together

Reply:
Regarding `CubeModel` vs. multi-integration, as far as I know, right now the only type of product `CubeModel` is used for is when you have 2-D data from multiple integrations, resulting in a 3-D stack. Other forms of pure 3-D data, such as IFU cubes, use their own unique data models, like `IFUCubeModel`. So right now there is a one-to-one correspondence between multi-integration and `CubeModel`, at least when you're past the ramp_fit stage (otherwise it's still 4-D) and you're before the spectral extraction stage (then you get into models like `MultiSpecModel`).

Reply:
And not sure how this factors in, but there will be (#2457 as soon as review is done), TSO-specific Level2 processing.

Reply:
Excellent!  Then we can punt on this issue because eventually TSO level 2 processing will have a separate pipeline?  We should resolve `is_tso()` at that point, and resolve it consistently.

Reply:
No, the TSO-specific level-2b processing won't involve a separate pipeline module, but it will use a separate config file ("calwebb_tso_spec2.cfg"), which by default has skip = True for the `resample_spec` step. So it might accomplish the same thing, at least when the pipeline is executed using the cfg file (won't help if someone just calls it directly in python without the cfg).

But again, this will only address exposures explicitly tagged as being a TSO mode and hence invoke (in the default DMS processing environment) the `calwebb_tso_spec2` config. The original problematic data set is not TSO and hence nothing would change up front for that case (would still run the regular `calwebb_spec2` pipeline and hence the `is_tso()` hook is still needed to skip `resample_spec`.

Reply:
I see.  So any further change required to this PR then?

Reply:
Just followed up in conversation with @jdavies-st.  As @philhodge and @hbushouse suggested, this should be a dimensionality check rather than relying on checking for `tso` which is not very explicit on what it is checking for.   Once that is updated, this can be merged.   A unit test should be added to check for that, and I'll open a separate issue that the `CubeModel` should be more descriptive.